source/common/shared_pool/shared_pool.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <thread>
+#include <type_traits>
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/singleton/instance.h"
+
+#include "common/common/assert.h"
+#include "common/common/non_copyable.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace SharedPool {
+
+/**
+ * Used to share objects that have the same content.
+ * control the life cycle of shared objects by reference counting
+ * Note: this class has only deleteObject for thread safety, and the other member methods are
+ * non-thread safe
+ */
+template <typename T, typename HashFunc = std::hash<T>,
+          class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
+class ObjectSharedPool : public Singleton::Instance,
+                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
+                         NonCopyable {
+public:
+  ObjectSharedPool(Event::Dispatcher& dispatcher)
+      : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
+  ~ObjectSharedPool() = default;
+
+  void deleteObject(const size_t hash_key) {
+    if (std::this_thread::get_id() == thread_id_) {
+      object_pool_.erase(hash_key);
+    } else {
+      auto this_shared_ptr = this->shared_from_this();
+      dispatcher_.post([hash_key, this_shared_ptr] { this_shared_ptr->deleteObject(hash_key); });
+    }
+  }
+
+  std::shared_ptr<T> getObject(const T& obj) {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    auto hashed_value = HashFunc{}(obj);
+    auto object_it = object_pool_.find(hashed_value);
+    if (object_it != object_pool_.end()) {
+      auto lock_object = object_it->second.lock();
+      if (lock_object) {
+        return lock_object;
+      }
+    }
+
+    auto this_shared_ptr = this->shared_from_this();
+    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
+      this_shared_ptr->deleteObject(hashed_value);
+      delete ptr;
```

Comment:
I *think* this can race due to the same hash value. Meaning, you delete the pointer here, but then you lock the weak pointer on the main thread and then you would cash with no internal value? I think you need to actually do the deletion on the main thread? Can you fix if necessary and add more comments/tests around this?

Reply:
I don't think there will be a race condition, because the object destructor and the object obtained through the weak_ptr are thread-safe, and when the object destructor is not finished, weak_ptr cannot get the real object through lock.

But I found another problem when an object is destructed, if it is not in the main thread, it will post a task to the main thread, and if the main thread gets an object with the same hash value through `getObject` at this time, it will cause a new weak_ptr to be stored in the map, and then execute `deleteObject` will delete the new weak_ptr.  I have fixed this issue and added the related unit test.

Reply:
https://en.cppreference.com/w/cpp/memory/weak_ptr/expired

> Equivalent to use_count() == 0. The destructor for the managed object may not yet have been called, but this object's destruction is imminent (or may have already happened).

Reply:
Can you add more comments here? I see that you switched the order of delete and delete object?

Reply:
There is no particular intention, just hope to release the ptr as soon as possible, which can avoid the ptr being exposed by mistake, resulting in some undefined behavior.

Reply:
Who would be accessing ptr? This seems pretty unclear to me. I think the comment should be further clarified with an example, as it's unclear what the concrete danger is here.

Reply:
example:

```
    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
      this_shared_ptr->sync().syncPoint(ObjectSharedPool<T>::ObjectDeleterEntry);
       // The error passes the ptr to a function that may not be executed immediately.
       delete ptr;
    });
```